### PR TITLE
Added a schema and a writer for the 2.0 NLU Parser

### LINKED ISSUE #6252: Add schema and writer for 2.0 NLU data
Part of #6129 

- Add YAML schema for the 2.0 YAML NLU data
- Add a NLU writer (will be used to convert MD to the 2.0 YAML)

### PR DESCRIPTION
**Pull request by degiz**
Closes #6252

**Proposed changes**:
- Added a schema and a writer for the 2.0 NLU format
- Extracted common parts of writers for both the MD and YAML
- Fixed a bug with synonyms not being extracted from the entities

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by degiz** [comment]
@federicotdn tagging you just in case, but the PR is not ready yet 😬

**Comment by degiz** [comment]
@tmbo @federicotdn asking review from either one of you, both are not necessary 😄

File: rasa/nlu/training_data/formats/rasa_yaml.py [hunk 1]
```diff
@@ -142,7 +170,7 @@ def _parse_training_examples(
         if isinstance(examples, list):
             example_strings = [
                 # pytype: disable=attribute-error
-                example.get(KEY_INTENT_TEXT, "")
+                example.get(KEY_INTENT_TEXT, "").strip("\n\r ")
```
Comment (tmbo) [inline rasa/nlu/training_data/formats/rasa_yaml.py:173]: why do we need the strip?

File: rasa/nlu/training_data/formats/rasa_yaml.py [hunk 2]
```diff
@@ -29,6 +47,8 @@
 
 MULTILINE_TRAINING_EXAMPLE_LEADING_SYMBOL = "-"
 
+NLU_SCHEMA_FILE = "nlu/schemas/nlu.yml"
```
Comment (tmbo) [inline rasa/nlu/training_data/formats/rasa_yaml.py:50]: would it make sense in the future to have a "rasa" schema file combining core, nlu, responses and domain (given that you can combine their contents in files)?

File: rasa/nlu/training_data/formats/rasa_yaml.py [hunk 3]
```diff
@@ -322,3 +350,108 @@ def is_yaml_nlu_file(filename: Text) -> bool:
...
         return False
+
+
+class RasaYAMLWriter(TrainingDataWriter):
+    """Writes training data into a file in a YAML format."""
+
+    def dumps(self, training_data: "TrainingData") -> Text:
+        """Turns TrainingData into a string."""
+        stream = StringIO()
+        self.dump(stream, training_data)
```
Comment (tmbo) [inline rasa/nlu/training_data/formats/rasa_yaml.py:361]: mhm `stream` isn't a filename (and not a `Text`) as stated in the below signature

File: rasa/nlu/training_data/formats/rasa_yaml.py [hunk 4]
```diff
@@ -322,3 +350,108 @@ def is_yaml_nlu_file(filename: Text) -> bool:
...
+
+        result = OrderedDict()
+        result[KEY_TRAINING_DATA_FORMAT_VERSION] = DoubleQuotedScalarString(
+            LATEST_TRAINING_DATA_FORMAT_VERSION
+        )
+        result[KEY_NLU] = nlu_items
+
+        io_utils.write_yaml(result, filename, True)
+
+    @staticmethod
```
Comment (tmbo) [inline rasa/nlu/training_data/formats/rasa_yaml.py:388]: should this rather be a classmethod, e.g. we could call`cls.process_training_examples_by_key`

File: rasa/nlu/training_data/formats/rasa_yaml.py [hunk 5]
```diff
@@ -322,3 +350,108 @@ def is_yaml_nlu_file(filename: Text) -> bool:
...
+        key_examples: Text,
+        example_extraction_predicate=lambda x: x,
+    ) -> List[OrderedDict]:
+        from ruamel.yaml.scalarstring import LiteralScalarString
+
+        result = []
+        for entity_key, examples in training_examples.items():
+
+            converted_examples = []
+            converted_examples += [
```
Comment (tmbo) [inline rasa/nlu/training_data/formats/rasa_yaml.py:445]:
```suggestion
            converted_examples = [
```

File: rasa/nlu/training_data/formats/rasa_yaml.py [hunk 6]
```diff
@@ -322,3 +350,108 @@ def is_yaml_nlu_file(filename: Text) -> bool:
...
+                TrainingDataWriter.generate_list_item(
+                    example_extraction_predicate(example).strip("\n\r ")
```
Comment (tmbo) [inline rasa/nlu/training_data/formats/rasa_yaml.py:447]: what if an example contains new lines?

File: rasa/core/training/interactive.py [hunk 7]
```diff
@@ -840,13 +843,21 @@ def _write_nlu_to_file(export_nlu_path: Text, events: List[Dict[Text, Any]]) ->
...
+        JSON_FILE_EXTENSION,
+    )
+
     guessed_format = loading.guess_format(export_path)
 
-    if guessed_format not in {MARKDOWN, RASA}:
-        if export_path.endswith(".json"):
+    if guessed_format not in {MARKDOWN, RASA, RASA_YAML}:
+        if export_path.endswith(JSON_FILE_EXTENSION):
             guessed_format = RASA
```
[the remark this answers is not in the document]
Comment (degiz) [reply rasa/core/training/interactive.py:856]: I don't think so, that's excplicitly the "extension" of the file as  opposed to the "format". The same way we have `YAML_FILE_EXTENSIONS`

File: rasa/nlu/training_data/formats/rasa_yaml.py (hunk 1 above)
Comment (degiz) [reply rasa/nlu/training_data/formats/rasa_yaml.py:173]: In testing phase I faced several problems with leading/trailing spaces and `\n` in the training examples, and found this as a safe solution to always take only "text" from the training examples/name of the intents.
At least I haven't seen the trailing white space as a vital part of a training data.

File: rasa/nlu/training_data/formats/rasa_yaml.py (hunk 2 above)
Comment (degiz) [reply rasa/nlu/training_data/formats/rasa_yaml.py:50]: Yes, I'm also thinking about that.. Will definitely do in the future.

File: rasa/nlu/training_data/formats/rasa_yaml.py (hunk 3 above)
Comment (degiz) [reply rasa/nlu/training_data/formats/rasa_yaml.py:361]: True! Fixed that.

File: rasa/nlu/training_data/formats/rasa_yaml.py (hunk 4 above)
Comment (degiz) [reply rasa/nlu/training_data/formats/rasa_yaml.py:388]: makes sense.

File: rasa/nlu/training_data/formats/rasa_yaml.py (hunk 6 above)
Comment (degiz) [reply rasa/nlu/training_data/formats/rasa_yaml.py:447]: That's a good question! The current implementation of reader/writer is done with an assumption that every line should start with `-` and be a separate training example, e.g. in the reader we're using the `splitlines()` all over.
I guess this has to be noted in the docs and possibly be changed before the release.

File: rasa/nlu/training_data/formats/readerwriter.py [hunk 8]
```diff
@@ -33,6 +46,87 @@ def dumps(self, training_data: "TrainingData") -> Text:
...
+
+        entity_dict = {
+            ENTITY_ATTRIBUTE_TYPE: entity_type,
+            ENTITY_ATTRIBUTE_ROLE: entity_role,
+            ENTITY_ATTRIBUTE_GROUP: entity_group,
+            ENTITY_ATTRIBUTE_VALUE: entity_value,
+        }
+        entity_dict = {k: v for k, v in entity_dict.items() if v is not None}
+
+        return f"[{entity_text}]{json.dumps(entity_dict)}"
```
[the remark this answers is not in the document]
Comment (degiz) [reply rasa/nlu/training_data/formats/readerwriter.py:128]: that's a very good point, thanks!

File: rasa/nlu/training_data/formats/rasa_yaml.py (hunk 6 above)
Comment (tmbo) [reply rasa/nlu/training_data/formats/rasa_yaml.py:447]: that sounds good, I think in the future we can make the splitting more advanced, and e.g. only treat lines starting with `-` as a new example. until then we might be able to escape the newlines? (using a literal `\n`)

File: rasa/nlu/training_data/formats/rasa_yaml.py (hunk 1 above)
Comment (tmbo) [reply rasa/nlu/training_data/formats/rasa_yaml.py:173]: makes sense :+1:

File: rasa/nlu/training_data/formats/rasa_yaml.py (hunk 6 above)
Comment (degiz) [reply rasa/nlu/training_data/formats/rasa_yaml.py]: Yes, the `\n` works for now!

File: rasa/utils/io.py [hunk 9]
```diff
@@ -12,18 +12,24 @@
...
+from ruamel.yaml import RoundTripRepresenter
 
 from rasa.constants import ENV_LOG_LEVEL, DEFAULT_LOG_LEVEL, YAML_VERSION
 
 if TYPE_CHECKING:
     from prompt_toolkit.validation import Validator
 
 DEFAULT_ENCODING = "utf-8"
+ESCAPE_DCT = {"\b": "\\b", "\f": "\\f", "\n": "\\n", "\r": "\\r", "\t": "\\t"}
+ESCAPE = re.compile(r"[\b\f\n\r\t]")
```
Comment (tmbo) [inline rasa/utils/io.py:29]: can we use the dict in L28 to create the regex?

File: tests/nlu/training_data/formats/test_rasa_yaml.py [hunk 10]
```diff
@@ -1,43 +1,109 @@
...
-    f"   metadata:\n"
-    f"   examples:\n"
-    f"    - text: |\n"
-    f"        how much CO2 will that use?\n"
-    f"      metadata:\n"
-    f"        sentiment: positive\n"
-    f"    - text: |\n"
-    f'        how much carbon will a one way flight from [new york]{{"entity": "city", "role": "from"}} to california produce?'
+MINIMAL_VALID_EXAMPLE = (
+    f"version: '{LATEST_TRAINING_DATA_FORMAT_VERSION}'\n" f"nlu:\n" f"stories:"
```
Comment (tmbo) [inline tests/nlu/training_data/formats/test_rasa_yaml.py:58]:
```suggestion
    f"version: '{LATEST_TRAINING_DATA_FORMAT_VERSION}'\n nlu:\n stories:"
```

**Pull request merged**